Refector Python bot to Java

skyegallup opened the pull request:
Closes #2. This is a full re-implementation of [the Python bot](https://github.com/codeRIT/lehrbot). It is not an identical copy; most notably, it features [server-unique topic names](https://github.com/codeRIT/lehrbot/issues/3) and [uses "topic" instead of "class"](https://github.com/codeRIT/lehrbot/issues/1). I will copy the other GitHub issues over from the old repository after this PR is closed.

Commit:
Added basic "on ready" code

Commit:
Added Topic model

Commit:
Added Server model

Commit:
Added command handling framework code

Commit:
Added topic management commands

Commit:
Added queue interaction commands

Commit:
Implemented help command

Commit:
Implemented clear command

Commit:
Cleanup unused imports

skyegallup commented:
I should be available for most of the weekend if anyone needs me to host the bot or provide the API key as part of their code review.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+            channel.sendMessage(String.format(
+                    "%s Topic \"%s\" does not exist.",
+                    member.getAsMention(),
+                    topic)).complete();
```
@Madrugaur fill in `.queue` info

Madrugaur replied:
.complete() is blocking so it will force the executing thread to wait for completion before continuing. If you have functionality that requires the action of sending a message be synchronous then you should use .complete(). In this case, I think kicking the task onto a queue, using .queue(), to be executed when the JDA gets around to it is a better solution. That way the bot isn't blocked and can continue to process commands.

Madrugaur replied:
@colbygallup

skyegallup replied:
I have swapped all of the `.complete()` calls to `.queue()`. I think that should be fine; I can't think of any calls that need to be synchronous.

Madrugaur commented on src/main/java/entities/Server.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.List;
+
+/**
+ * An object that provides Topic CRUD methods for a Guild.
```
What is CRUD?

Madrugaur commented on src/main/java/entities/Server.java:
```diff
@@ -0,0 +1,71 @@
+     * @param guild The guild to build from
+     * @return A Server with all topics added
+     */
+    public static Server load(Guild guild) {
```
I usually put all my static methods at the bottom of a class. This is just an organization thing but I think it makes it easier to read.

skyegallup replied:
I put this static method below the constructor because it acts similar to a constructor, but I'll shift it downwards to make it consistent with authbot's styling.

Madrugaur commented on src/main/java/entities/Server.java:
```diff
@@ -0,0 +1,71 @@
+        guild.createRole()
+                .setName("Topic | " + topic.getName())
+                .setMentionable(true)
+                .complete();
```
Again, maybe think about replacing with .queue(). In this context, blocking the thread might be a good idea because then you can handle any errors that come from this operation failing. But you would have to weigh that out.

Madrugaur commented on src/main/java/entities/Server.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.List;
+
+/**
+ * An object that provides Topic CRUD methods for a Guild.
```
What is CRUD? Could you expand on what that means in the doc string? Also you might want to add an '@author' tag.

skyegallup replied:
CRUD = create, read, update, delete. It's usually used in the context of public-facing REST APIs, which this class isn't. I'll reword it to make more sense.

Madrugaur commented on src/main/java/entities/Server.java:
```diff
@@ -0,0 +1,71 @@
+     */
+    public void deleteTopic(Topic topic) {
+        guild.getRolesByName("Topic | " + topic.getName(), true)
+                .forEach((role -> role.delete().complete()));
```
Consider using .queue() instead of .complete()

Madrugaur commented on src/main/java/entities/Topic.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class Topic {
+    private final String name;
+    private final ArrayList<Member> queue = new ArrayList<>();
```
ArrayLists have terrible time complexity for addition and removal of elements. Since you're using it as a queue, I would switch to a data structure with a better time complexity. Java's LinkedList class can be used the exact same way as an ArrayList but with a constant time complexity for removals and additions. Additionally, if you wanted to stick to the name you could use one of the implementations of a queue that Java Collections Framework has.

Easiest: Just use a LinkedList
More inline with the name: use a Queue

Madrugaur replied:
For reference, ArrayLists have an O(n) time complexity as opposed to a LinkedList's O(1)

skyegallup replied:
This is a good catch. I'll switch this to LinkedList, since that class implements the Queue interface.

Commit:
API interactions are now async

Madrugaur commented on src/main/java/launcher/Lehrry.java:
```diff
@@ -1,7 +1,23 @@
-        System.out.println("hey!");
+        try {
+            JDA jda = JDABuilder.createDefault(System.getenv("LEHRRY_TOKEN"))
+                    .setChunkingFilter(ChunkingFilter.ALL)
```
If you want the member list to be updated with all of the members of the guild, you need to add .setMemberCachePolicy(MemberCachePolicy.ALL) or something like that. You can check Auther for the exact code

skyegallup replied:
I don't believe my code accesses the guild's member list at any point, so this would only add unnecessary event receipts.

Madrugaur commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+     */
+    private static boolean checkAdmin(Member member, TextChannel channel) {
+        if (!member.hasPermission(Permission.ADMINISTRATOR)) {
+            channel.sendMessage(member.getAsMention() + " You must have administrator permission to run this command.").complete();
```
Again with .queue()

Madrugaur commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+        channel.sendMessage(String.format(
+                "%s Topic role \"%s\" has been created.",
+                member.getAsMention(),
+                args[0])).complete();
```
queue()

Madrugaur commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+        channel.sendMessage(String.format(
+                "%s Topic role \"%s\" has been deleted.",
+                member.getAsMention(),
+                args[0])).complete();
```
Im so sorry... queue()

Madrugaur commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+        channel.sendMessage(String.format(
+                "%s List of topics:\n%s",
+                member.getAsMention(),
+                topicList)).complete();
```
😬 queue()

Madrugaur commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+            channel.sendMessage(String.format(
+                    "%s has left the \"%s\" queue.",
+                    member.getAsMention(),
+                    args[0])).complete();
```
plz dont hate me, but queue()

Madrugaur commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+            channel.sendMessage(String.format(
+                    "%s has joined the \"%s\" queue.",
+                    member.getAsMention(),
+                    args[0])).complete();
```
ok im stopping now, just find all the rest

Madrugaur commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+                mentee.getAsMention())).complete();
+    }
+
+    private void showqueue(Member member, TextChannel channel, Server server, String[] args) {
```
Maybe make it so this command can only be called from a specific channel. It feels a little undesirable to just post the names of the people waiting in a chat anyone can access.

skyegallup replied:
You make a good point, but this falls outside of the scope of this PR. I will create a new issue for it after this PR gets merged.

Madrugaur replied:
Dope, just popped into my head at that moment

peterkos replied:
We can def re-evaluate the exact functionality sometime this week and at next week's meeting!

Madrugaur commented:
@colbygallup

Commit:
Moved Server.load down for styling

Commit:
Reworded Server docstring

Commit:
Switched ArrayList to LinkedList

skyegallup commented:
@Madrugaur

Madrugaur reviewed (approved):
This all looks good to me, thanks for putting up with all my comments lmao

peterkos commented on src/main/java/entities/Server.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class Server {
+    private final Guild guild;
+    private final ArrayList<Topic> topics = new ArrayList<>();
```
Just a note: might be cool to have separate categories of topics: a set for langs, a set for frameworks, a set for design help, a set for "how to build X" (where to start building backend, app, etc.)

skyegallup replied:
This sounds like a good idea. I will add it as a separate issue.

skyegallup replied:
Included in #12

peterkos commented on src/main/java/entities/Server.java:
```diff
@@ -0,0 +1,72 @@
+     * @param guild The guild to build from
+     * @return A Server with all topics added
+     */
+    public static Server load(Guild guild) {
```
fActOriEs? in MY jAVa??!

skyegallup replied:
I tend to use factory methods a lot in object-oriented programming, but I don't really know if that's a good or bad thing. Should this remain as-is, or should this be converted to a `Server(Guild guild)` constructor?

peterkos replied:
I personally prefer non-factories (there's nothing wrong with them!). Although they're much more common in Java than other languages, I think with the scale of our project it'd be okay to keep consistent and just use a constructor.

skyegallup replied:
Alright, I moved the factory code into the constructor.

peterkos commented on src/main/java/entities/Topic.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class Topic {
+    private final String name;
+    private final LinkedList<Member> queue = new LinkedList<>();
```
`final`?

skyegallup replied:
In Java, the `final` modifier prevents a variable from being reassigned or assigned after instance construction. It prevents us, for example, from calling `this.queue = new LinkedList<>()` in any instance method, since it would be better to empty the existing list than to create a new one. It's not required, but it follows the idea that developers should be given as few methods as possible to modify an object's state, similar to why we use getter/setter methods instead of making fields public.

peterkos replied:
Gotcha! 
After using a language like Swift the concept of getters/setters stops being useful -- the theoretical of "a developer Might change an object they already have read/write access too" doesn't apply to self-contained projects (i.e., not an API), and at that point a Lot of boilerplate goes away :)

skyegallup replied:
I do agree that we don't *need* to use as many restrictions because we aren't developing with the intent of interaction with other people's code. However, restrictions like these also keep us from making mistakes as we move forward (e.g. accidentally changing the name attribute of a Topic class instead of recreating the object with a new name), so I feel that stuff like this does help us avoid issues in the long run.

peterkos commented on src/main/java/entities/Topic.java:
```diff
@@ -0,0 +1,71 @@
+     * @return The Members in this queue
+     */
+    public Member[] getMembersInQueue() {
+        return queue.toArray(new Member[0]);
```
Wouldn't this be
```Java
return queue.toArray(new Member[queue.size()]);
```
?

skyegallup replied:
That's what I thought, but IntelliJ gave me a warning saying that this was a better alternative. I think `Collection.toArray()` automatically creates an array of the proper size if the existing one is too small, so this works.

skyegallup replied:
IntelliJ's explanation:

> There are two styles to convert a collection to an array: either using a pre-sized array (like c.toArray(new String[c.size()])) or using an empty array (like c.toArray(new String[0]). In older Java versions using pre-sized array was recommended, as the reflection call which is necessary to create an array of proper size was quite slow. However since late updates of OpenJDK 6 this call was intrinsified, making the performance of the empty array version the same and sometimes even better, compared to the pre-sized version. Also passing pre-sized array is dangerous for a concurrent or synchronized collection as a data race is possible between the size and toArray call which may result in extra nulls at the end of the array, if the collection was concurrently shrunk during the operation. This inspection allows to follow the uniform style: either using an empty array (which is recommended in modern Java) or using a pre-sized array (which might be faster in older Java versions or non-HotSpot based JVMs).

peterkos replied:
I love these paragraph explanations! Could you add a small comment summing that up for future ref?

skyegallup replied:
Yeah, I'll add a comment here.

peterkos commented on src/main/java/launcher/Lehrry.java:
```diff
@@ -1,7 +1,23 @@
+
+import javax.security.auth.login.LoginException;
+
 public class Lehrry {
```
Qu'est-ce que c'est `Lehrry`?

skyegallup replied:
If `authbot` -> `Auther`, then `lehrbot` -> `Lehrry`. I can change it to `Mentorbot` if you'd prefer, though.

peterkos replied:
That'd be great!

peterkos commented on src/main/java/launcher/Lehrry.java:
```diff
@@ -1,7 +1,23 @@
+                    .addEventListeners(new MainEventListener())
+                    .build();
+        } catch (LoginException ex) {
+            ex.printStackTrace();
```
Unified error handling is something we should keep in mind going forward! (is ok for this PR)

skyegallup replied:
Yep, I'll create an issue about better error handling later. This reminds me that we also need to make an issue about input validation, since I don't think user input is actually validated for any of the commands.

skyegallup replied:
Now that I've looked at the code again, it appears that most input *is* validated, but no errors are displayed to the user. I'll still make an issue for it.

skyegallup replied:
Included in #10.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+     * @return Contains the Topic, if it exists
+     */
+    private static Optional<Topic> checkTopicExists(Member member, TextChannel channel, Server server, String topic) {
+        Optional<Topic> topicOptional = Arrays.stream(server.getTopics())
```
I haven't worked with optionals in Java, only in Swift -- is it convention to explicitly name it `xOptional`?

skyegallup replied:
My research (over the past ~5 minutes) has shown that there aren't any strong naming conventions, but that `optionalX` and `optX` are the prevalent formats. I will switch this variable name to `optionalTopic`.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+        // pick the correct method to call
+        CommandHandler commandHandler;
+        switch (tokens[0].substring(1)) {
+            case "help" -> commandHandler = this::help;
```
nit: could the `->` all be aligned?

peterkos replied:
(I have a plugin in SublimeText that does this if that seems like a strange request; I'm used to doing it for long things like this)

skyegallup replied:
Yeah, I can work on aligning these. I don't have a plugin for it, but it shouldn't take too long to do manually right now. If it does, I'll try to find one for IntelliJ.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+            default -> commandHandler = this::unknownCommand;
+        }
+        commandHandler.handle(
+                Objects.requireNonNull(event.getMember()),
```
Will this throw exception if not null? Will that catch somewhere higher and send an informative message to user?

skyegallup replied:
This throws `NullPointerException` if `event.getMember()` is null. I only added this on IntelliJ's suggestion; I think `event.getMember()` only returns `null` for a Discord-generated join/leave message, but those will probably never be detected as a bot command.

skyegallup replied:
It is not caught anywhere else, but like I said, I don't think this can ever be caused by an actual user's message, so I don't see any point in showing an error message.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+        embedBuilder.setDescription("Possible commands:");
+        embedBuilder.setColor(0xE57D25);
+
+        embedBuilder.addField("$queue <topic>", "Add yourself to a queue.", false);
```
It might be a good idea to factor these into a static class so we can change any of the verbiage / view all possible commands at once, something like `Message.addSelfToQueue` for this one, etc.

skyegallup replied:
That's a good idea, but I feel like something that large would best be done as a separate issue.

skyegallup replied:
Included in #9.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+
+        embedBuilder.addField("$queue <topic>", "Add yourself to a queue.", false);
+        embedBuilder.addField("$showqueue <topic>", "Show the people currently in queue.", false);
+        embedBuilder.addField("$showtopics", "List all topics.", false);
```
Also what is the last bool parameter here?

skyegallup replied:
The last bool parameter determines whether the field is shown inline or not. In [this example](https://preview.redd.it/e7v6dvh5k3n41.png?width=607&format=png&auto=webp&s=37af4696cf03c623dc41684ec10f254511d09ee0), the first two fields are inline (`true`), while the second two are not inline (`false`).

peterkos replied:
Ah kk!
I wish every language had named parameters...

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+        embedBuilder.addField("$queue <topic>", "Add yourself to a queue.", false);
+        embedBuilder.addField("$showqueue <topic>", "Show the people currently in queue.", false);
+        embedBuilder.addField("$showtopics", "List all topics.", false);
+        embedBuilder.addField("$ready <topic> (admin only)", "Retrieve the next person from the queue.", false);
```
Is there a way to only show these commands to users with the admin role?

peterkos replied:
(Out of scope but just to get some discussion)

skyegallup replied:
There should be, but it has a lot of overlap with #5. I will bundle these suggestions into that issue.

skyegallup replied:
Actually, on second thought, this would need to show three versions of the help list: one for normal users, one for mentors, and one for admins. In that case, this feature would be out of scope of #5. I'll make it into a separate issue.

skyegallup replied:
Included in #8.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    private void deletetopic(Member member, TextChannel channel, Server server, String[] args) {
+        if (!checkAdmin(member, channel)) return;
```
I'd suggest commenting early return conditions just to make sure the intent is there

skyegallup replied:
What do you mean? A comment like `// do not allow non-admins to run command` on line 133?

peterkos replied:
Yep! there are other instances when it's directly stated as code but I prefer making exit conditions as clear as possible

skyegallup replied:
Alright, I'll add comments where applicable.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+        if (!checkAdmin(member, channel)) return;
+
+        Arrays.stream(server.getTopics())                   // loop over all topics...
+                .filter(t -> t.getName().equals(args[0]))   // ...if this topic's name is args[0]...
```
What is `args[0]`?

skyegallup replied:
`args[0]` is the first parameter provided to the command. For example, if a user calls `$deletetopic Java`, `args == ["Java"]`.

peterkos replied:
I ask because I known it's the first arg of the command, but _semantically_? Like, is it the `channelToDelete`, that kinda thing

skyegallup replied:
I can add an extra variable to make it read better semantically. Something like `string topicName = args[0];` on line 134?

peterkos replied:
That'd be great! I just think it helps the code clarity

skyegallup replied:
Added the extra variables where necessary

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+                args[0])).queue();
+    }
+
+    private void deletetopic(Member member, TextChannel channel, Server server, String[] args) {
```
camelCase?

skyegallup replied:
I was trying to name the per-command functions to be identical to how they are represented to the user. In this case, the actual command is `$deletetopic <topic>`, so I wanted the function name to match. However, I suppose that it doesn't matter since the user can't see the code anyway, so I can swap these to camelCase.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+                topicList)).queue();
+    }
+
+    private void queue(Member member, TextChannel channel, Server server, String[] args) {
```
A bit of documentation for the method would be great, even just something like:
```java
/** Queues a user for a specific topic. Operates per-server. **/
```
Or we can go full javadoc to be consistent w/ `authbot-hm` -- thoughts?

peterkos replied:
(Adding docs should be in sep issue btw; this PR is a bit large as-is lol)

skyegallup replied:
Yep. External documentation is already discussed in #4; should internal documentation like this be a separate issue, or get merged into that one?

peterkos replied:
I'll come back to this

peterkos replied:
Let's do it as part of #4

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    private void queue(Member member, TextChannel channel, Server server, String[] args) {
+        Optional<Topic> topicOptional = checkTopicExists(member, channel, server, args[0]);
```
Could `args[0]` be refactored into a variable just to make it a bit more semantic in the code?

skyegallup replied:
I left `args` as an array to provide support for future commands that may require multiple parameters (e.g. `$removeuser <user mention> <topic>` to remove a user from one topic's queue). Switching `args` to a single variable thus only makes sense if we're 100% certain that no future commands will use multiple parameters.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+    private void unknownCommand(Member member, TextChannel channel, Server server, String[] args) {
+        channel.sendMessage(member.getAsMention() + " Command does not exist! Try $help for a list of valid commands.").queue();
+    }
+}
```
Newline @ EOF? Although GitHub sometimes doesn't show it

skyegallup replied:
You're right, I forgot to add a newline there. I'll make sure to fix that.

skyegallup replied:
Nope, never mind on this one as well. There is definitely a newline here on my end, but GitHub isn't showing it. I'm not sure if there's anything I can do to fix that.

peterkos replied:
No worries; GH complains usually but other times it just shows it, other times it doesn't. 
@natfriedman hire me so i can fix bug thanks

peterkos replied:
Oh wow the @ actually worked

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+                .filter(t -> t.getName().equalsIgnoreCase(topic))
+                .findFirst();
+        if (topicOptional.isEmpty()) {
+            channel.sendMessage(String.format(
```
I'd definitely recommend refactoring error messages into a static class -- see [here](https://github.com/codeRIT/brickhack-mobile-ios/blob/master/BrickHack-Mobile/AlertMessage.swift) for iOS, although this seems much easier to present given it's text based 😅

skyegallup replied:
What benefits would this have? I'm not saying it's a bad idea, just that I don't fully understand why that would be better.

skyegallup replied:
Never mind; I just saw [this comment](https://github.com/codeRIT/mentorbot/pull/3/files/f6d09b4f9f565d44adc98382deaaa87b345c53a8#r515673643) and it makes sense now.

skyegallup replied:
Also included in #9.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+                Objects.requireNonNull(event.getMember()),
+                event.getChannel(),
+                server,
+                Arrays.copyOfRange(tokens, 1, tokens.length));
```
nit: Might be a bit cleaner if at the beginning of the method the args were extracted to variables to be a bit more semantic

skyegallup replied:
Yep, will fix this one as well.

peterkos commented on src/main/java/listeners/MainEventListener.java:
```diff
@@ -0,0 +1,232 @@
+        Optional<Topic> topicOptional = checkTopicExists(member, channel, server, args[0]);
+        if (topicOptional.isEmpty()) return;
+
+        Topic topic = topicOptional.get();
```
Just want to say I really appreciate your use of optionals! #safeprogramming

peterkos reviewed (changes requested):
Great work!

A chunk of questions I have are not explicitly related to this code so feel free to write off as out-of-scope. 

As mentioned in one of the comments, I haven't tested (or really considered) much of the actual functionality; here I'm more focused on the code quality itself. Later this week, or at next week's meeting, we'll take a closer look at what we intend to have and make a full feature list on Notion just to make sure the bot's on the path we want!

I'll also check in the Directors meeting on Tuesday to see if there's any functionality they'd like to see here.

Commit:
Renamed command functions to camelCase

Commit:
Moved command handler arguments to separate variables

Commit:
Prettified the command switch

Commit:
Rename `topicOptional` to `optionalTopic`

skyegallup commented:
@peterkos Finished looking through your review. Fixed some, debated others, moved a few to separate issues. Let me know if these are all fine or if further changes are needed.

Commit:
Moved Server factory code into constructor

Commit:
Renamed Lehrry to Mentorbot

Commit:
Added comments to early returns

Commit:
IntelliJ's weird suggestions, explained

skyegallup commented:
@peterkos Added most of the other suggestions. Just a couple more things to check, then we can merge this.

peterkos reviewed (approved):
I'll approve; just needs this final change and is good!

Commit:
Added more variables for readability
